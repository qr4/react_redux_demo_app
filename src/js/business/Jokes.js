// @flow

export class Jokes {
    +listOfJokes: string[];

    constructor(initialListOfJokes: string[] = []) {
        this.listOfJokes = initialListOfJokes;
    }

    addNewJoke(joke: string): Jokes {
        return new Jokes([...this.listOfJokes, joke]);
    }

    removeJokeAtPosition(position: number): Jokes {
        if (position < 0 || this.listOfJokes.length <= position) {
            return this;
        }
        return new Jokes([...this.listOfJokes.slice(0, position), ...this.listOfJokes.slice(position + 1)]);
    }
}
