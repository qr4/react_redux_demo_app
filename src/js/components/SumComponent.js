// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';

import type { State } from '../state/Reducer';
import { Jokes } from '../business/Jokes';

export class DisconnectedSumComponent extends Component<{ jokes: Jokes }> {
    render() {
        const sum = this.props.jokes.listOfJokes.reduce((red: number, elem: string) => red + elem.length, 0);
        return <div className="well test-class-sum">{'Sum of all characters: ' + sum}</div>;
    }
}

const mapStateToProps = (state: State): { jokes: Jokes } => ({ jokes: state.jokes });

export const SumComponent = connect(mapStateToProps)(DisconnectedSumComponent);
