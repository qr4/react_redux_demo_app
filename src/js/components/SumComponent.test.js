// @flow

import React from 'react';
import { shallow } from 'enzyme';

import { DisconnectedSumComponent } from './SumComponent';
import { Jokes } from '../business/Jokes';

describe('SumComponent', () => {
    it('[snapshot] it renders correctly', () => {
        const app = shallow(<DisconnectedSumComponent jokes={new Jokes()} />);
        expect(app).toMatchSnapshot();
    });

    it('computes the sum of characters correctly', () => {
        const jokes = new Jokes().addNewJoke('a').addNewJoke('123');
        const app = shallow(<DisconnectedSumComponent jokes={jokes} />);

        const elem = app.find('.test-class-sum');
        expect(elem.contains('Sum of all characters: 4')).toBe(true);
    });
});
