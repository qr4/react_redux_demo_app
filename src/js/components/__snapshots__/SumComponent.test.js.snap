// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SumComponent [snapshot] it renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DisconnectedSumComponent
    jokes={
        Jokes {
            "listOfJokes": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Sum of all characters: 0",
      "className": "well test-class-sum",
    },
    "ref": null,
    "rendered": "Sum of all characters: 0",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Sum of all characters: 0",
        "className": "well test-class-sum",
      },
      "ref": null,
      "rendered": "Sum of all characters: 0",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
